### ISSUE TITLE
Openssl 3.0.5 Atom C3958 Hardware QAT

### ISSUE THREAD

**Issue by jimthedj65** [intel/QAT_Engine#218]
Hi All,

I have installed all the prerequisites for my Ubuntu server 22.04 edge box Lanner NCA1516a box, which has an Atom C3000 (C3958) and OpenSSL 3.0.5 hardware only support, and I am sure QAT_Engine supports it, I would appreciate being corrected.

I ran `./autogen.sh
./configure --with-qat_hw_dir=/home/path_to/QAT  --enable-qat_provider (I used the Qat provider to support the openssl 3.0)`
All seems to compile fine with no errors
`checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a race-free mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking how to print strings... printf
checking whether make supports the include directive... yes (GNU style)
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether gcc accepts -g... yes
checking for gcc option to enable C11 features... none needed
checking whether gcc understands -c and -o together... yes
checking dependency style of gcc... gcc3
checking for a sed that does not truncate output... /usr/bin/sed
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for fgrep... /usr/bin/grep -F
checking for ld used by gcc... /usr/bin/ld
checking if the linker (/usr/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 1572864
checking how to convert x86_64-pc-linux-gnu file names to x86_64-pc-linux-gnu format... func_convert_file_noop
checking how to convert x86_64-pc-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /usr/bin/ld option to reload object files... -r
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ar... ar
checking for archiver @FILE support... @
checking for strip... strip
checking for ranlib... ranlib
checking command to parse /usr/bin/nm -B output from gcc object... ok
checking for sysroot... no
checking for a working dd... /usr/bin/dd
checking how to truncate binary pipes... /usr/bin/dd bs=4096 count=1
checking for mt... mt
checking if mt is a manifest tool... no
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... yes
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking whether ln -s works... yes
checking for g++... g++
checking whether the compiler supports GNU C++... yes
checking whether g++ accepts -g... yes
checking for g++ option to enable C++11 features... none needed
checking dependency style of g++... gcc3
checking how to run the C++ preprocessor... g++ -E
checking for ld used by g++... /usr/bin/ld -m elf_x86_64
checking if the linker (/usr/bin/ld -m elf_x86_64) is GNU ld... yes
checking whether the g++ linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking for g++ option to produce PIC... -fPIC -DPIC
checking if g++ PIC flag -fPIC -DPIC works... yes
checking if g++ static flag -static works... yes
checking if g++ supports -c -o file.o... yes
checking if g++ supports -c -o file.o... (cached) yes
checking whether the g++ linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking dynamic linker characteristics... (cached) GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking for gcc... (cached) gcc
checking whether the compiler supports GNU C... (cached) yes
checking whether gcc accepts -g... (cached) yes
checking for gcc option to enable C11 features... (cached) none needed
checking whether gcc understands -c and -o together... (cached) yes
checking dependency style of gcc... (cached) gcc3
checking for /home/path_to/QAT/include/qat/icp_sal_versions.h... no
checking for pkg-config... /usr/bin/pkg-config
configure: Build QAT Provider against OpenSSL 3.0
configure: Install engine/provider in the system OpenSSL Path
configure: Enabling build for QAT HW acceleration
configure: Build with existing error files without regenerating from OpenSSL source
configure: Accelerating RSA to Hardware
configure: Accelerating X25519/X448 to Hardware
configure: Accelerating ECDH to Hardware
configure: Accelerating ECDSA to Hardware
configure: Accelerating DSA to Hardware
configure: Accelerating DH to Hardware
configure: Accelerating PRF to Hardware
configure: Not Accelerating HKDF to Hardware
configure: Accelerating CIPHERS to Hardware
configure: Not Accelerating SM2 to Software
configure: Not Accelerating SM3 to Software
configure: Not Accelerating SHA3 to Hardware
configure: Not Accelerating CHACHA-POLY to Hardware
configure: Not Accelerating small packet CIPHERS to Hardware
configure: Lenstra attack protection enabled (default).
configure: Disabling Engine Debug Messages
configure: Disabling Engine Warning Messages
configure: Disabling Memory Driver Debug Messages
configure: Disabling Memory Driver Warning Messages
configure: Auto Engine init on fork enabled (default).
configure: Disabling QAT SW Heuristic Timeout (default)
configure: Disabling CPU Cycle Counts
configure: Disabling QAT hardware set instance for thread
configure: Compiling for USDM
configure: Using default engine id "qatengine"
checking for ldd... yes
checking for head... yes
configure: glibc version >= 2.23. glibc AVX-512 memcpy optimisation supported.
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating config.h
config.status: config.h is unchanged
config.status: executing depfiles commands
config.status: executing libtool commands`

`make
make  all-am
make[1]: Entering directory '/home/path_to/QAT_Engine'
make[1]: Leaving directory '/home/path_to/QAT_Engine'`

```
make install
make[1]: Entering directory '/home/path_to/QAT_Engine'
 /usr/bin/mkdir -p '/usr/lib/x86_64-linux-gnu/ossl-modules'
 /bin/bash ./libtool   --mode=install /usr/bin/install -c   qatengine.la '/usr/lib/x86_64-linux-gnu/ossl-modules'
libtool: install: /usr/bin/install -c .libs/qatengine.so /usr/lib/x86_64-linux-gnu/ossl-modules/qatengine.so
libtool: install: /usr/bin/install -c .libs/qatengine.lai /usr/lib/x86_64-linux-gnu/ossl-modules/qatengine.la
libtool: finish: PATH="/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/sbin" ldconfig -n /usr/lib/x86_64-linux-gnu/ossl-modules
```
----------------------------------------------------------------------
Libraries have been installed in:
   /usr/lib/x86_64-linux-gnu/ossl-modules

If you ever happen to want to link against installed libraries
in a given directory, LIBDIR, you must either use libtool, and
specify the full pathname of the library, or use the '-LLIBDIR'
flag during linking and do at least one of the following:
   - add LIBDIR to the 'LD_LIBRARY_PATH' environment variable
     during execution
   - add LIBDIR to the 'LD_RUN_PATH' environment variable
     during linking
   - use the '-Wl,-rpath -Wl,LIBDIR' linker flag
   - have your system administrator add LIBDIR to '/etc/ld.so.conf'

See any operating system documentation about shared libraries for
more information, such as the ld(1) and ld.so(8) manual pages.
----------------------------------------------------------------------
make[1]: Nothing to be done for 'install-data-am'.

make[1]: Leaving directory '/home/path_to/QAT_Engine'`

I added include /usr/lib/x86_64-linux-gnu/ossl-modules to my /etc/ld.so.conf 

The errors I get when running the tests are:
 openssl engine -t -c -v qatengine
409710B7A67F0000:error:12800067:DSO support routines:dlfcn_load:could not load the shared library:../crypto/dso/dso_dlfcn.c:118:filename(/usr/lib/x86_64-linux-gnu/engines-3/qatengine.so): /usr/lib/x86_64-linux-gnu/engines-3/qatengine.so: undefined symbol: qat_num_instances
409710B7A67F0000:error:12800067:DSO support routines:DSO_load:could not load the shared library:../crypto/dso/dso_lib.c:152:
409710B7A67F0000:error:13000084:engine routines:dynamic_load:dso not found:../crypto/engine/eng_dyn.c:422:
409710B7A67F0000:error:13000074:engine routines:ENGINE_by_id:no such engine:../crypto/engine/eng_list.c:430:id=qatengine

I have noticed that my libraries for openssl 3.0 are located in /usr/lib/x86_64-linux-gnu/ossl-modules but when I run openssl engine -t -c -v qatengine

It points to /usr/lib/x86_64-linux-gnu/engines-3/

My previous build was on a 8360Y with openssl 1.1.1k and I have confused myself somewhere with Atom and openssl 3.0. 

I obviously have something confused in my environment and can't see it for looking any assistance would be appreciated.

**Comment by Yogaraj-Alamenda** [comment]
@jimthedj65 From your configure  you have enabled provider but trying to use engine from OpenSSL 3.0. If you would like to use engine just remove E--enable-qat_providerE from the build configure. If you prefer to use qatprovider use option -provider qatprovider in the OpenSSL Speed as mentioned in [here](https://github.com/intel/QAT_Engine/blob/master/docs/qat_common.md#openssl-30-provider-support)

**Comment by jimthedj65** [comment]
thank you for the response, I got sidetracked on my dev build.  I will report back.

**Comment by jimthedj65** [comment]
I have reverted this to a new post https://github.com/intel/QAT_Engine/issues/246

**Comment by jimthedj65** [comment]
was resolved by building with intree driver see #246
